templatekit/repo.py

```diff
@@ -0,0 +1,277 @@
+"""Template repository APIs.
+"""
+
+__all__ = ('Repo', 'FileTemplate', 'ProjectTemplate', 'BaseTemplate')
+
+import os
+import itertools
+import logging
+
+
+class Repo(object):
+    """Template repository.
+
+    Parameters
+    ----------
+    root : `str`
+        Path to the root directory of the template repository. Use ``'.'``
+        as the current working directory.
+    """
+
+    def __init__(self, root):
+        super().__init__()
+        self._log = logging.getLogger(__name__)
+        self.root = root
+
+    @classmethod
+    def discover_repo(cls, dirname='.'):
+        """Create a Repo instance by discovering the template repo's
+        root directory.
+
+        Parameters
+        ----------
+        dirname : `str`
+            Relative or absolute path of a directory. This directory should
+            either be the root of a template repository, or a subdirectory
+            of the template repository.
+
+        Returns
+        -------
+        repo : `Repo`
+            The Repo instance corresponding to the given directory.
+
+        Raises
+        ------
+        OSError
+            Raised if ``dirname`` is not, or is not contained by, a
+            recognizable templates repository.
+        """
+        original_dirname = dirname
+        dirname = os.path.abspath(dirname)
+
+        while dirname != '/':
+            if cls._is_repo_dir(dirname):
+                return cls(dirname)
+
+            # Repeat with the parent directory
+            dirname = os.path.split(dirname)[0]
+
+        message = ('The directory {0!r} is not contained by a recognizable '
+                   'template repository.')
+        raise OSError(message.format(original_dirname))
+
+    @staticmethod
+    def _is_repo_dir(dirname):
+        if not os.path.isdir(os.path.join(dirname, 'file_templates')):
+            return False
+
+        if not os.path.isdir(os.path.join(dirname, 'project_templates')):
+            return False
+
+        return True
+
+    def __repr__(self):
+        return 'Repo({0!r})'.format(self.root)
+
+    def __str__(self):
+        return '{0!r}\nProject templates: {1!s}\nFile templates: {2!s}'.format(
+            self,
+            ', '.join([t.name for t in self.iter_project_templates()]),
+            ', '.join([t.name for t in self.iter_file_templates()])
+        )
+
+    def __iter__(self):
+        """Iterate over the names of all templates in the repository.
+
+        ``__contains__`` delegates to this method.
+        """
+        for template in self.iter_templates():
+            yield template.name
+
+    def __getitem__(self, key):
```

Comment:
Couldn't you subclass `dict` to get this functionality?

Edit: I guess not as you are storing two lists internally.